File: ts/packages/core/src/models/Webhooks.ts
/**
 * @fileoverview The `composio.webhooks` namespace: project webhook
 * subscriptions (outbound event delivery) and webhook endpoints (inbound
 * per-toolkit ingress URLs).
 *
 * @module Webhooks
 */
import ComposioClient from '@composio/client';
import { z } from 'zod/v3';
import {
  DefaultWebhookSubscriptionEvents,
  WebhookSubscription,
  WebhookVersions,
} from '../types/triggers.types';
import {
  WebhookEndpoint,
  WebhookEndpointCreateParams,
  WebhookEndpointCreateParamsSchema,
  WebhookEndpointListParams,
  WebhookEndpointListParamsSchema,
  WebhookEndpointListResponse,
  WebhookEndpointReplaceParams,
  WebhookEndpointReplaceParamsSchema,
  WebhookEndpointUpdateParams,
  WebhookEndpointUpdateParamsSchema,
  WebhookSubscriptionEventTypesResponse,
  WebhookSubscriptionDeleteResponse,
  WebhookSubscriptionListParams,
  WebhookSubscriptionListParamsSchema,
  WebhookSubscriptionListResponse,
  WebhookSubscriptionRotateSecretResponse,
  WebhookSubscriptionSetParams,
  WebhookSubscriptionSetParamsSchema,
  WebhookSubscriptionUpdateParams,
  WebhookSubscriptionUpdateParamsSchema,
} from '../types/webhooks.types';
import { ComposioRequestOptions } from '../types/requestOptions.types';
import { ValidationError } from '../errors/ValidationErrors';
import { telemetry } from '../telemetry/Telemetry';
import { withCancellation } from '../utils/cancellation';
import {
  transformWebhookEndpoint,
  transformWebhookEndpointListResponse,
  transformWebhookSubscriptionEventTypesResponse,
  transformWebhookSubscription,
  transformWebhookSubscriptionDeleteResponse,
  transformWebhookSubscriptionListResponse,
  transformWebhookSubscriptionRotateSecretResponse,
} from '../utils/transformers/webhooks';

/**
 * The only part of the `list({ limit: 1 })` probe the upsert needs: the id of
 * the first (and, since the API allows one subscription per project, only)
 * subscription. Parsed at the boundary so a malformed payload never reaches
 * the update path.
 */
const FirstWebhookSubscriptionIdSchema = z
  .object({
    items: z.array(z.object({ id: z.string().min(1) }).passthrough()).optional(),
  })
  .passthrough()
  .transform(response => response.items?.[0]?.id);

/**
 * Create-or-update the project webhook subscription.
 *
 * Shared by `composio.webhooks.subscriptions.set()` and the legacy
 * `composio.triggers.setWebhookSubscription()` so both issue the exact same
 * call sequence: `webhookSubscriptions.list({ limit: 1 })`, then
 * `webhookSubscriptions.update(id, body)` when a subscription exists,
 * otherwise `webhookSubscriptions.create(body)`.
 *
 * @internal
 */
export async function upsertWebhookSubscription(
  client: ComposioClient,
  params: WebhookSubscriptionSetParams,
  requestOptions?: ComposioRequestOptions
): Promise<WebhookSubscription> {
  const parsedParams = WebhookSubscriptionSetParamsSchema.safeParse(params);

  if (!parsedParams.success) {
    throw new ValidationError(`Invalid parameters passed to set webhook subscription`, {
      cause: parsedParams.error,
    });
  }

  const body = {
    webhook_url: parsedParams.data.webhookUrl,
    enabled_events: parsedParams.data.enabledEvents ?? [...DefaultWebhookSubscriptionEvents],
    version: parsedParams.data.version ?? WebhookVersions.V3,
  };

  const existing = await withCancellation(
    () => client.webhookSubscriptions.list({ limit: 1 }, requestOptions),
    requestOptions?.signal
  );
  const parsedExisting = FirstWebhookSubscriptionIdSchema.safeParse(existing);
  if (!parsedExisting.success) {
    throw new ValidationError('Failed to parse webhook subscription list response', {
      cause: parsedExisting.error,
    });
  }
  const subscriptionId = parsedExisting.data;

  const subscription = await withCancellation(
    () =>
      subscriptionId
        ? client.webhookSubscriptions.update(subscriptionId, body, requestOptions)
        : client.webhookSubscriptions.create(body, requestOptions),
    requestOptions?.signal
  );

  return transformWebhookSubscription(subscription);
}

/**
 * `composio.webhooks.subscriptions` — the project's outbound webhook
 * subscription. The API allows one subscription per project; every response
 * includes the signing `secret` used to verify deliveries with
 * `composio.triggers.verifyWebhook()` / `composio.triggers.parse()`.
 */
export class WebhookSubscriptions {
  private client: ComposioClient;

  constructor(client: ComposioClient) {
    this.client = client;
    telemetry.instrument(this, 'WebhookSubscriptions');
  }

  /**
   * List the project's webhook subscriptions.
   *
   * @param {WebhookSubscriptionListParams} [query] - Pagination options
   * @returns {Promise<WebhookSubscriptionListResponse>} Paginated subscriptions
   *
   * @example
   * ```typescript
   * const { items } = await composio.webhooks.subscriptions.list();
   * console.log(items[0]?.webhookUrl);
   * ```
   */
  async list(
    query?: WebhookSubscriptionListParams,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookSubscriptionListResponse> {
    const parsedQuery = WebhookSubscriptionListParamsSchema.safeParse(query ?? {});
    if (!parsedQuery.success) {
      throw new ValidationError('Failed to parse webhook subscription list query', {
        cause: parsedQuery.error,
      });
    }
    const listParams = {
      limit: parsedQuery.data.limit,
      cursor: parsedQuery.data.cursor,
    };
    const result = await withCancellation(
      () => this.client.webhookSubscriptions.list(listParams, requestOptions),
      requestOptions?.signal
    );
    return transformWebhookSubscriptionListResponse(result);
  }

  /**
   * Retrieve a webhook subscription by id.
   *
   * @param {string} id - The subscription id
   * @returns {Promise<WebhookSubscription>} The subscription, including its signing secret
   *
   * @example
   * ```typescript
   * const subscription = await composio.webhooks.subscriptions.get('whs_abc123');
   * console.log(subscription.enabledEvents);
   * ```
   */
  async get(id: string, requestOptions?: ComposioRequestOptions): Promise<WebhookSubscription> {
    const result = await withCancellation(
      () => this.client.webhookSubscriptions.retrieve(id, requestOptions),
      requestOptions?.signal
    );
    return transformWebhookSubscription(result);
  }

  /**
   * Create or update the project webhook subscription.
   *
   * If a subscription already exists it is updated in place, otherwise a new
   * one is created. By default this subscribes to V3 trigger message events.
   *
   * @param {WebhookSubscriptionSetParams} params - `webhookUrl`, optional `enabledEvents` and `version`
   * @returns {Promise<WebhookSubscription>} The resulting subscription
   * @throws {ValidationError} If the params fail validation
   *
   * @example
   * ```typescript
   * const subscription = await composio.webhooks.subscriptions.set({
   *   webhookUrl: `${APP_URL}/webhooks/composio`,
   *   enabledEvents: ['composio.trigger.message', 'composio.connected_account.expired'],
   * });
   * process.env.COMPOSIO_WEBHOOK_SECRET = subscription.secret;
   * ```
   */
  async set(
    params: WebhookSubscriptionSetParams,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookSubscription> {
    return upsertWebhookSubscription(this.client, params, requestOptions);
  }

  /**
   * Update an existing webhook subscription. Only the provided fields change.
   *
   * @param {string} id - The subscription id
   * @param {WebhookSubscriptionUpdateParams} params - Fields to update
   * @returns {Promise<WebhookSubscription>} The updated subscription
   * @throws {ValidationError} If the params fail validation
   *
   * @example
   * ```typescript
   * await composio.webhooks.subscriptions.update('whs_abc123', {
   *   enabledEvents: ['composio.trigger.message'],
   * });
   * ```
   */
  async update(
    id: string,
    params: WebhookSubscriptionUpdateParams,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookSubscription> {
    const parsedParams = WebhookSubscriptionUpdateParamsSchema.safeParse(params);
    if (!parsedParams.success) {
      throw new ValidationError('Failed to parse webhook subscription update params', {
        cause: parsedParams.error,
      });
    }
    const body = {
      webhook_url: parsedParams.data.webhookUrl,
      enabled_events: parsedParams.data.enabledEvents,
      version: parsedParams.data.version,
    };
    const result = await withCancellation(
      () => this.client.webhookSubscriptions.update(id, body, requestOptions),
      requestOptions?.signal
    );
    return transformWebhookSubscription(result);
  }

  /**
   * Delete a webhook subscription. Deliveries stop immediately.
   *
   * @param {string} id - The subscription id
   * @returns {Promise<WebhookSubscriptionDeleteResponse>} `{ success }`
   *
   * @example
   * ```typescript
   * await composio.webhooks.subscriptions.delete('whs_abc123');
   * ```
   */
  async delete(
    id: string,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookSubscriptionDeleteResponse> {
    const result = await withCancellation(
      () => this.client.webhookSubscriptions.delete(id, requestOptions),
      requestOptions?.signal
    );
    return transformWebhookSubscriptionDeleteResponse(result);
  }

  /**
   * Rotate the signing secret of a webhook subscription. The previous secret
   * stops validating, so update your verifier with the returned `secret`.
   *
   * @param {string} id - The subscription id
   * @returns {Promise<WebhookSubscriptionRotateSecretResponse>} `{ id, secret }`
   *
   * @example
   * ```typescript
   * const { secret } = await composio.webhooks.subscriptions.rotateSecret('whs_abc123');
   * ```
   */
  async rotateSecret(
    id: string,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookSubscriptionRotateSecretResponse> {
    const result = await withCancellation(
      () => this.client.webhookSubscriptions.rotateSecret(id, requestOptions),
      requestOptions?.signal
    );
    return transformWebhookSubscriptionRotateSecretResponse(result);
  }

  /**
   * List the event types a subscription can enable, with the payload
   * versions each one supports.
   *
   * @returns {Promise<WebhookSubscriptionEventTypesResponse>} The available event types
   *
   * @example
   * ```typescript
   * const { items } = await composio.webhooks.subscriptions.listEventTypes();
   * console.log(items.map(item => item.eventType));
   * ```
   */
  async listEventTypes(
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookSubscriptionEventTypesResponse> {
    const result = await withCancellation(
      () => this.client.webhookSubscriptions.listEventTypes(requestOptions),
      requestOptions?.signal
    );
    return transformWebhookSubscriptionEventTypesResponse(result);
  }
}

/**
 * `composio.webhooks.endpoints` — inbound webhook endpoints. Each endpoint is
 * the ingress URL Composio exposes for one toolkit + OAuth app pair, which you
 * register with the third-party provider so it can push events to Composio.
 */
export class WebhookEndpoints {
  private client: ComposioClient;

  constructor(client: ComposioClient) {
    this.client = client;
    telemetry.instrument(this, 'WebhookEndpoints');
  }

  /**
   * List webhook endpoints, optionally filtered by toolkit.
   *
   * @param {WebhookEndpointListParams} [query] - Optional `toolkitSlug` filter
   * @returns {Promise<WebhookEndpointListResponse>} The endpoints
   *
   * @example
   * ```typescript
   * const { items } = await composio.webhooks.endpoints.list({ toolkitSlug: 'slack' });
   * ```
   */
  async list(
    query?: WebhookEndpointListParams,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookEndpointListResponse> {
    const parsedQuery = WebhookEndpointListParamsSchema.safeParse(query ?? {});
    if (!parsedQuery.success) {
      throw new ValidationError('Failed to parse webhook endpoint list query', {
        cause: parsedQuery.error,
      });
    }
    const listParams = { toolkit_slug: parsedQuery.data.toolkitSlug };
    const result = await withCancellation(
      () => this.client.webhookEndpoints.list(listParams, requestOptions),
      requestOptions?.signal
    );
    return transformWebhookEndpointListResponse(result);
  }

  /**
   * Retrieve a webhook endpoint by its nano id.
   *
   * @param {string} nanoId - The endpoint nano id (e.g. `we_abc123`)
   * @returns {Promise<WebhookEndpoint>} The endpoint, with secret setup values masked
   *
   * @example
   * ```typescript
   * const endpoint = await composio.webhooks.endpoints.get('we_abc123');
   * console.log(endpoint.webhookUrl);
   * ```
   */
  async get(nanoId: string, requestOptions?: ComposioRequestOptions): Promise<WebhookEndpoint> {
    const result = await withCancellation(
      () => this.client.webhookEndpoints.retrieve(nanoId, requestOptions),
      requestOptions?.signal
    );
    return transformWebhookEndpoint(result);
  }

  /**
   * Create a webhook endpoint for a toolkit + OAuth app pair. Returns the
   * existing endpoint when one already exists for that pair.
   *
   * @param {WebhookEndpointCreateParams} params - `toolkitSlug` and `clientId`
   * @returns {Promise<WebhookEndpoint>} The endpoint; register `webhookUrl` with the provider
   * @throws {ValidationError} If the params fail validation
   *
   * @example
   * ```typescript
   * const endpoint = await composio.webhooks.endpoints.create({
   *   toolkitSlug: 'slack',
   *   clientId: 'my-slack-app-client-id',
   * });
   * ```
   */
  async create(
    params: WebhookEndpointCreateParams,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookEndpoint> {
    const parsedParams = WebhookEndpointCreateParamsSchema.safeParse(params);
    if (!parsedParams.success) {
      throw new ValidationError('Failed to parse webhook endpoint create params', {
        cause: parsedParams.error,
      });
    }
    const body = {
      toolkit_slug: parsedParams.data.toolkitSlug,
      client_id: parsedParams.data.clientId,
    };
    const result = await withCancellation(
      () => this.client.webhookEndpoints.create(body, requestOptions),
      requestOptions?.signal
    );
    return transformWebhookEndpoint(result);
  }

  /**
   * Replace the setup fields of a webhook endpoint. All required setup fields
   * for the toolkit must be provided — this is the initial configuration.
   *
   * @param {string} nanoId - The endpoint nano id
   * @param {WebhookEndpointReplaceParams} params - `data` with every required setup field
   * @returns {Promise<WebhookEndpoint>} The configured endpoint
   * @throws {ValidationError} If the params fail validation
   *
   * @example
   * ```typescript
   * await composio.webhooks.endpoints.replace('we_abc123', {
   *   data: { signing_secret: process.env.SLACK_SIGNING_SECRET! },
   * });
   * ```
   */
  async replace(
    nanoId: string,
    params: WebhookEndpointReplaceParams,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookEndpoint> {
    const parsedParams = WebhookEndpointReplaceParamsSchema.safeParse(params);
    if (!parsedParams.success) {
      throw new ValidationError('Failed to parse webhook endpoint replace params', {
        cause: parsedParams.error,
      });
    }
    const result = await withCancellation(
      () =>
        this.client.webhookEndpoints.replace(
          nanoId,
          { data: parsedParams.data.data },
          requestOptions
        ),
      requestOptions?.signal
    );
    return transformWebhookEndpoint(result);
  }

  /**
   * Merge setup fields into a webhook endpoint. Omitted fields are preserved.
   *
   * @param {string} nanoId - The endpoint nano id
   * @param {WebhookEndpointUpdateParams} params - `data` with the fields to change
   * @returns {Promise<WebhookEndpoint>} The updated endpoint
   * @throws {ValidationError} If the params fail validation
   *
   * @example
   * ```typescript
   * await composio.webhooks.endpoints.update('we_abc123', {
   *   data: { signing_secret: 'rotated-secret' },
   * });
   * ```
   */
  async update(
    nanoId: string,
    params: WebhookEndpointUpdateParams,
    requestOptions?: ComposioRequestOptions
  ): Promise<WebhookEndpoint> {
    const parsedParams = WebhookEndpointUpdateParamsSchema.safeParse(params);
    if (!parsedParams.success) {
      throw new ValidationError('Failed to parse webhook endpoint update params', {
        cause: parsedParams.error,
      });
    }
    const result = await withCancellation(
      () =>
        this.client.webhookEndpoints.update(
          nanoId,
          { data: parsedParams.data.data },
          requestOptions
        ),
      requestOptions?.signal
    );
    return transformWebhookEndpoint(result);
  }
}

/**
 * `composio.webhooks` namespace.
 *
 * - `subscriptions`: the project's outbound webhook subscription (where
 *   Composio delivers trigger and lifecycle events).
 * - `endpoints`: inbound ingress URLs for third-party providers, one per
 *   toolkit + OAuth app pair.
 *
 * @example
 * ```typescript
 * const subscription = await composio.webhooks.subscriptions.set({
 *   webhookUrl: 'https://example.com/webhooks/composio',
 * });
 * const endpoints = await composio.webhooks.endpoints.list({ toolkitSlug: 'slack' });
 * ```
 */
export class Webhooks {
  /** Manage the project's outbound webhook subscription. */
  subscriptions: WebhookSubscriptions;
  /** Manage inbound webhook endpoints per toolkit + OAuth app. */
  endpoints: WebhookEndpoints;

  constructor(client: ComposioClient) {
    this.subscriptions = new WebhookSubscriptions(client);
    this.endpoints = new WebhookEndpoints(client);
    telemetry.instrument(this, 'Webhooks');
  }
}

File: python/composio/core/models/webhooks.py
"""
Webhooks module for the Composio SDK.

Exposes ``composio.webhooks.subscriptions`` (the project-level webhook
subscription used for event delivery) and ``composio.webhooks.endpoints``
(per toolkit + OAuth app inbound ingress URLs).
"""

from __future__ import annotations

import typing as t
from enum import Enum

import pydantic

from composio import exceptions
from composio.client import HttpClient
from composio.client.types import (
    webhook_endpoint_create_response,
    webhook_endpoint_list_response,
    webhook_endpoint_replace_response,
    webhook_endpoint_retrieve_response,
    webhook_endpoint_update_response,
    webhook_subscription_delete_response,
    webhook_subscription_list_event_types_response,
    webhook_subscription_list_response,
    webhook_subscription_retrieve_response,
    webhook_subscription_rotate_secret_response,
    webhook_subscription_update_response,
)
from composio.core.models.base import Resource
from composio.utils.pydantic import none_to_omit


class WebhookVersion(str, Enum):
    """Webhook payload version."""

    V1 = "V1"
    V2 = "V2"
    V3 = "V3"


WebhookVersionL = t.Literal["V1", "V2", "V3"]

_WEBHOOK_VERSIONS: t.Dict[str, WebhookVersionL] = {"V1": "V1", "V2": "V2", "V3": "V3"}


def _webhook_version(version: t.Union[WebhookVersion, str]) -> WebhookVersionL:
    """Narrow a caller-supplied version to the literal the client accepts."""
    value = version.value if isinstance(version, WebhookVersion) else version
    literal = _WEBHOOK_VERSIONS.get(value)
    if literal is None:
        raise exceptions.ValidationError(
            f"unsupported webhook version {value!r}; expected one of V1, V2, V3"
        )
    return literal


class WebhookSubscription(t.TypedDict, total=False):
    """Webhook subscription returned by the Composio API."""

    id: str
    webhook_url: str
    version: str
    enabled_events: t.List[str]
    secret: str
    created_at: str
    updated_at: str


DEFAULT_WEBHOOK_SUBSCRIPTION_EVENTS = ("composio.trigger.message",)


def normalize_webhook_subscription(raw: object) -> WebhookSubscription:
    """Build a typed :class:`WebhookSubscription` from an API response.

    Accepts either the client's pydantic response models or a plain dict.
    Maps explicitly (accepting either snake_case or camelCase wire keys)
    instead of ``cast``-ing the raw object, so the returned dict always
    matches the declared shape and a shift in the wire format surfaces as a
    normalized field rather than a ``KeyError`` at the call site.
    """
    if isinstance(raw, pydantic.BaseModel):
        # The client builds response models without validation, so enum
        # fields hold plain strings; silence pydantic's serializer warnings.
        data: t.Dict[str, t.Any] = raw.model_dump(mode="json", warnings=False)
    elif isinstance(raw, dict):
        data = raw
    else:
        data = {}

    def _first_str(*keys: str) -> t.Optional[str]:
        for key in keys:
            value = data.get(key)
            if isinstance(value, str) and value:
                return value
        return None

    def _str_list(*keys: str) -> t.List[str]:
        for key in keys:
            value = data.get(key)
            if isinstance(value, list):
                return [item for item in value if isinstance(item, str)]
        return []

    result: WebhookSubscription = {
        "id": _first_str("id") or "",
        "webhook_url": _first_str("webhook_url", "webhookUrl") or "",
        "version": _first_str("version") or WebhookVersion.V3.value,
        "enabled_events": _str_list("enabled_events", "enabledEvents"),
    }
    secret = _first_str("secret")
    if secret is not None:
        result["secret"] = secret
    created_at = _first_str("created_at", "createdAt")
    if created_at is not None:
        result["created_at"] = created_at
    updated_at = _first_str("updated_at", "updatedAt")
    if updated_at is not None:
        result["updated_at"] = updated_at
    return result


def first_webhook_subscription_id(response: object) -> t.Optional[str]:
    """Return the id of the first subscription in a list response.

    Accepts the client's typed list response (``.items`` of models with an
    ``.id``) as well as a plain dict of the same shape.
    """
    items = (
        response.get("items")
        if isinstance(response, dict)
        else getattr(response, "items", None)
    )
    if not isinstance(items, list) or len(items) == 0:
        return None

    first = items[0]
    subscription_id = (
        first.get("id") if isinstance(first, dict) else getattr(first, "id", None)
    )
    return subscription_id if isinstance(subscription_id, str) else None


class WebhookSubscriptions(Resource):
    """Manage the project webhook subscription.

    The API allows one subscription per project; the signing secret is
    returned in responses.
    """

    def list(
        self,
        *,
        limit: t.Optional[float] = None,
        cursor: t.Optional[str] = None,
    ) -> webhook_subscription_list_response.WebhookSubscriptionListResponse:
        """
        List webhook subscriptions for the project.

        :param limit: Maximum number of subscriptions to return.
        :param cursor: Pagination cursor from a previous response.
        :return: Paginated list of subscriptions.

        Example:
            subscriptions = composio.webhooks.subscriptions.list()
        """
        return self._client.webhook_subscriptions.list(
            limit=none_to_omit(limit),
            cursor=none_to_omit(cursor),
        )

    def get(
        self, id: str
    ) -> webhook_subscription_retrieve_response.WebhookSubscriptionRetrieveResponse:
        """
        Retrieve a webhook subscription by id.

        :param id: The subscription id.
        :return: The subscription, including its signing secret.

        Example:
            subscription = composio.webhooks.subscriptions.get("sub_123")
        """
        return self._client.webhook_subscriptions.retrieve(id)

    def set(
        self,
        *,
        webhook_url: str,
        enabled_events: t.Optional[t.Sequence[str]] = None,
        version: t.Union[WebhookVersion, str] = WebhookVersion.V3,
    ) -> WebhookSubscription:
        """
        Create or update the project webhook subscription used for webhook delivery.

        If a subscription already exists, the first subscription is updated. Otherwise a
        new subscription is created. By default this subscribes to V3 trigger message
        events.

        :param webhook_url: The URL Composio delivers events to.
        :param enabled_events: Event types to subscribe to. Defaults to
            ``composio.trigger.message``.
        :param version: Webhook payload version. Defaults to ``V3``.
        :return: The normalized subscription.

        Example:
            composio.webhooks.subscriptions.set(
                webhook_url=f"{APP_URL}/webhooks/composio",
            )
        """
        if not webhook_url:
            raise exceptions.ValidationError("please provide a valid `webhook_url`")

        events = list(
            DEFAULT_WEBHOOK_SUBSCRIPTION_EVENTS
            if enabled_events is None
            else enabled_events
        )
        if len(events) == 0:
            raise exceptions.ValidationError(
                "please provide at least one enabled event"
            )

        version_value = _webhook_version(version)

        existing = self._client.webhook_subscriptions.list(limit=1)
        subscription_id = first_webhook_subscription_id(existing)

        if subscription_id:
            return normalize_webhook_subscription(
                self._client.webhook_subscriptions.update(
                    subscription_id,
                    webhook_url=webhook_url,
                    enabled_events=events,
                    version=version_value,
                )
            )

        return normalize_webhook_subscription(
            self._client.webhook_subscriptions.create(
                webhook_url=webhook_url,
                enabled_events=events,
                version=version_value,
            )
        )

    def update(
        self,
        id: str,
        *,
        webhook_url: t.Optional[str] = None,
        enabled_events: t.Optional[t.Sequence[str]] = None,
        version: t.Union[WebhookVersion, str, None] = None,
    ) -> webhook_subscription_update_response.WebhookSubscriptionUpdateResponse:
        """
        Update an existing webhook subscription. Omitted fields are preserved.

        :param id: The subscription id.
        :param webhook_url: New delivery URL.
        :param enabled_events: New list of subscribed event types.
        :param version: New webhook payload version.
        :return: The updated subscription.

        Example:
            composio.webhooks.subscriptions.update(
                "sub_123",
                enabled_events=["composio.trigger.message"],
            )
        """
        version_value = None if version is None else _webhook_version(version)
        return self._client.webhook_subscriptions.update(
            id,
            webhook_url=none_to_omit(webhook_url),
            enabled_events=none_to_omit(
                list(enabled_events) if enabled_events is not None else None
            ),
            version=none_to_omit(version_value),
        )

    def delete(
        self, id: str
    ) -> webhook_subscription_delete_response.WebhookSubscriptionDeleteResponse:
        """
        Delete a webhook subscription.

        :param id: The subscription id.
        :return: The deletion response.

        Example:
            composio.webhooks.subscriptions.delete("sub_123")
        """
        return self._client.webhook_subscriptions.delete(id)

    def rotate_secret(
        self, id: str
    ) -> webhook_subscription_rotate_secret_response.WebhookSubscriptionRotateSecretResponse:
        """
        Rotate the signing secret of a webhook subscription.

        :param id: The subscription id.
        :return: The subscription with its new signing secret.

        Example:
            rotated = composio.webhooks.subscriptions.rotate_secret("sub_123")
            new_secret = rotated.secret
        """
        return self._client.webhook_subscriptions.rotate_secret(id)

    def list_event_types(
        self,
    ) -> webhook_subscription_list_event_types_response.WebhookSubscriptionListEventTypesResponse:
        """
        List the event types a subscription can be enabled for.

        :return: The available event types.

        Example:
            event_types = composio.webhooks.subscriptions.list_event_types()
        """
        return self._client.webhook_subscriptions.list_event_types()


class WebhookEndpoints(Resource):
    """Manage inbound webhook endpoints (per toolkit + OAuth app ingress URLs)."""

    def list(
        self,
        *,
        toolkit_slug: t.Optional[str] = None,
    ) -> webhook_endpoint_list_response.WebhookEndpointListResponse:
        """
        List webhook endpoints, optionally filtered by toolkit.

        :param toolkit_slug: Only return endpoints for this toolkit.
        :return: The list of endpoints.

        Example:
            endpoints = composio.webhooks.endpoints.list(toolkit_slug="github")
        """
        return self._client.webhook_endpoints.list(
            toolkit_slug=none_to_omit(toolkit_slug),
        )

    def get(
        self, nano_id: str
    ) -> webhook_endpoint_retrieve_response.WebhookEndpointRetrieveResponse:
        """
        Retrieve a webhook endpoint by id.

        :param nano_id: The endpoint id.
        :return: The endpoint.

        Example:
            endpoint = composio.webhooks.endpoints.get("we_123")
        """
        return self._client.webhook_endpoints.retrieve(nano_id)

    def create(
        self,
        *,
        toolkit_slug: str,
        client_id: str,
    ) -> webhook_endpoint_create_response.WebhookEndpointCreateResponse:
        """
        Create a webhook endpoint for a toolkit and OAuth app.

        :param toolkit_slug: The toolkit the endpoint receives events for.
        :param client_id: The OAuth app client id the endpoint is bound to.
        :return: The created endpoint, including its ingress ``webhook_url``.

        Example:
            endpoint = composio.webhooks.endpoints.create(
                toolkit_slug="github",
                client_id="Iv1.abc123",
            )
        """
        return self._client.webhook_endpoints.create(
            toolkit_slug=toolkit_slug,
            client_id=client_id,
        )

    def replace(
        self,
        nano_id: str,
        *,
        data: t.Dict[str, t.Any],
    ) -> webhook_endpoint_replace_response.WebhookEndpointReplaceResponse:
        """
        Replace the stored data of a webhook endpoint (full replacement).

        :param nano_id: The endpoint id.
        :param data: The complete new endpoint data.
        :return: The updated endpoint.

        Example:
            composio.webhooks.endpoints.replace(
                "we_123",
                data={"secret": "new-signing-secret"},
            )
        """
        return self._client.webhook_endpoints.replace(nano_id, data=data)

    def update(
        self,
        nano_id: str,
        *,
        data: t.Dict[str, t.Any],
    ) -> webhook_endpoint_update_response.WebhookEndpointUpdateResponse:
        """
        Merge new data into a webhook endpoint (partial update).

        :param nano_id: The endpoint id.
        :param data: The fields to merge into the endpoint data.
        :return: The updated endpoint.

        Example:
            composio.webhooks.endpoints.update(
                "we_123",
                data={"secret": "new-signing-secret"},
            )
        """
        return self._client.webhook_endpoints.update(nano_id, data=data)


class Webhooks(Resource):
    """Webhooks namespace, exposed as ``composio.webhooks``."""

    subscriptions: WebhookSubscriptions
    """Project webhook subscription used for event delivery."""

    endpoints: WebhookEndpoints
    """Inbound webhook endpoints (per toolkit + OAuth app)."""

    def __init__(self, client: HttpClient):
        super().__init__(client)
        self.subscriptions = WebhookSubscriptions(client)
        self.endpoints = WebhookEndpoints(client)
